#!/bin/bash
/BiO/Share/Tools/gatk-4.6.1.0/gatk ApplyVQSR --java-options "-XX:+UseSerialGC -Xmx55g" --reference /BiO/Share/Tools/gatk-bundle/hg38/Homo_sapiens_assembly38.fasta --variant /BiO/Teach/Standard-Pipeline/04_Germline_short_variant_discovery/C001_TN.rc.snp.vcf --output /BiO/Teach/Standard-Pipeline/04_Germline_short_variant_discovery/C001_TN.rc.indel.vcf -mode INDEL --recal-file /BiO/Teach/Standard-Pipeline/04_Germline_short_variant_discovery/C001_TN.indel.recal --tranches-file /BiO/Teach/Standard-Pipeline/04_Germline_short_variant_discovery/C001_TN.indel.tranches --truth-sensitivity-filter-level 99.9 --create-output-variant-index true